### ISSUE TITLE
Chapter 0 - SFTTrainer does not apply chat template leading to broken LLM output

### ISSUE THREAD

**Issue by Alzter** [dvgodoy/FineTuningLLMs#5]
# System Info

- Platform: Linux-6.6.72-x86_64-with-glibc2.40
- Python version: 3.12.8
- PyTorch version: 2.5.1+cu124
- CUDA device(s): NVIDIA GeForce RTX 2080 Ti, NVIDIA GeForce RTX 2080 Ti, NVIDIA GeForce RTX 2080 Ti, NVIDIA GeForce RTX 2080 Ti
- Transformers version: 4.46.2
- Accelerate version: 1.1.0
- Accelerate config: not found
- Datasets version: 2.21.0
- HF Hub version: 0.26.2
- TRL version: 0.15.2
- bitsandbytes version: 0.44.1
- DeepSpeed version: not installed
- Diffusers version: not installed
- Liger-Kernel version: not installed
- LLM-Blender version: not installed
- OpenAI version: not installed
- PEFT version: 0.13.2

## OS
```bash
    Operating System: NixOS 24.11 (Vicuna)            
         CPE OS Name: cpe:/o:nixos:nixos:24.11
              Kernel: Linux 6.6.72
        Architecture: x86-64
     Hardware Vendor: ASUSTeK COMPUTER INC.
      Hardware Model: WS X299 SAGE
    Firmware Version: 0402
```

<details>
<summary><h2>Python Libraries</h4></summary>

```bash
Package                   Version
------------------------- ---------------
accelerate                1.1.0
aiodns                    3.2.0
aiohappyeyeballs          2.4.2
aiohttp                   3.10.10
aiosignal                 1.3.1
anyio                     4.6.2
argon2-cffi               23.1.0
argon2-cffi-bindings      21.2.0
arrow                     1.3.0
asttokens                 2.4.1
async_generator           1.10
async-lru                 2.0.4
async-timeout             4.0.3
attrs                     24.2.0
babel                     2.16.0
bcrypt                    4.2.0
beautifulsoup4            4.12.3
bitsandbytes              0.44.1
bleach                    6.1.0
Brotli                    1.1.0
brotlicffi                1.1.0.0
certifi                   2024.8.30
cffi                      1.17.1
chardet                   5.2.0
charset-normalizer        3.3.2
comm                      0.2.2
contourpy                 1.3.0
cryptography              43.0.1
cycler                    0.12.1
datasets                  2.21.0
decorator                 5.1.1
defusedxml                0.8.0rc2
dill                      0.3.9
executing                 2.1.0
fastjsonschema            2.19.1
filelock                  3.16.1
fonttools                 4.54.1
fqdn                      1.5.1
frozenlist                1.4.1
fsspec                    2024.3.0
future                    1.0.0
h11                       0.14.0
html5lib                  1.1
httpcore                  1.0.6
httpx                     0.27.2
huggingface_hub           0.26.2
idna                      3.10
ipykernel                 6.29.5
ipython                   8.29.0
ipywidgets                8.1.5
isoduration               20.11.0
jedi                      0.19.1
Jinja2                    3.1.5
json5                     0.9.25
jsonpointer               3.0.0
jsonschema                4.23.0
jsonschema-specifications 2023.12.1
jupyter                   1.1.1
jupyter_client            8.6.3
jupyter-console           6.6.3
jupyter_core              5.7.2
jupyter-events            0.10.0
jupyter-lsp               2.2.5
jupyter_server            2.14.2
jupyter_server_terminals  0.5.3
jupyterlab                4.2.5
jupyterlab_pygments       0.3.0
jupyterlab_server         2.27.3
jupyterlab_widgets        3.0.13
kiwisolver                1.4.7
lit                       18.1.8
markdown-it-py            3.0.0
MarkupSafe                3.0.2
matplotlib                3.9.2
matplotlib-inline         0.1.7
mdurl                     0.1.2
mistune                   3.0.2
mpmath                    1.3.0
multidict                 6.1.0
multiprocess              0.70.16
nbclient                  0.10.0
nbconvert                 7.16.4
nbformat                  5.10.4
nest_asyncio              1.6.0
networkx                  3.3
notebook                  7.2.2
notebook_shim             0.2.4
numpy                     1.26.4
overrides                 7.7.0
packaging                 24.1
pandas                    2.2.3
pandocfilters             1.5.1
paramiko                  3.5.0
parso                     0.8.4
peft                      0.13.2
pexpect                   4.9.0
pillow                    11.0.0
pip                       24.0
platformdirs              4.3.6
prometheus_client         0.21.0
prompt_toolkit            3.0.48
protobuf                  5.28.3
psutil                    6.0.0
ptyprocess                0.7.0
pure_eval                 0.2.2
pyarrow                   18.0.0
pycares                   4.4.0
pycparser                 2.22
Pygments                  2.18.0
PyNaCl                    1.5.0
pyparsing                 3.1.4
pyspnego                  0.11.1
python-dateutil           2.9.0.post0
python-json-logger        2.0.7
pytz                      2024.2
PyYAML                    6.0.2
pyzmq                     26.2.0
referencing               0.35.1
regex                     2024.9.11
requests                  2.32.3
responses                 0.25.3
rfc3339_validator         0.1.4
rfc3986_validator         0.1.1
rich                      13.8.1
rpds-py                   0.18.1
safetensors               0.4.5
scipy                     1.14.1
Send2Trash                1.8.3
sentencepiece             0.2.0
setuptools                75.1.1.post0
six                       1.16.0
smbprotocol               1.14.0
sniffio                   1.3.1
soupsieve                 2.6
stack_data                0.6.3
sympy                     1.13.3
terminado                 0.18.1
tinycss2                  1.3.0
tokenizers                0.20.3
toml                      0.10.2
torch                     2.5.1+cu124
torchaudio                2.5.1+cu124
torchvision               0.20.1+cu124
tornado                   6.4.2
tqdm                      4.66.5
traitlets                 5.14.3
transformers              4.46.2
triton                    3.1.0
trl                       0.15.2
types-python-dateutil     2.9.0.20240906
types-PyYAML              6.0.12.20240917
types-toml                0.10.8.20240310
typing_extensions         4.12.2
tzdata                    2024.2
uri-template              1.3.0
urllib3                   2.2.3
wcwidth                   0.2.13
webcolors                 24.8.0
webencodings              0.5.1
websocket-client          1.8.0
widgetsnbextension        4.0.13
xxhash                    3.5.0
yarl                      1.13.1
```
</details>

# Issue
The fine-tuned LLM in Chapter 0 **no longer works** because `SFTTrainer` no longer formats the training dataset in the chat template when calling `SFTTrainer.__init__()`.

![Image](https://github.com/user-attachments/assets/9df800b3-2b47-48c6-ac3e-07816fc789b0)

# Steps to Reproduce
1. Run all cells in "Chapter 0" using the following libraries:

<table>
<tr><th>Library</th><th>Version</th></tr>
<tr><td>torch</td><td>2.5.1+cu124</td></tr>
<tr><td>trl</td><td>0.15.2</td></tr>
<tr><td>datasets</td><td>2.21.0</td></tr>
<tr><td>transformers</td><td>4.46.2</td></tr>
<tr><td>accelerate</td><td>1.1.0</td></tr>
<tr><td>bitsandbytes</td><td>0.44.1</td></tr>
</table>

# Expected Output
When querying the LLM:
```python
"<|user|> The Force is strong in you!<|end|><|assistant|> Strong in you, the Force is!<|end|>"
```

# Actual Output
When querying the LLM, the model doesn't stop generating nonsense:
```python
"<|user|> The Force is strong in you!<|end|><|assistant|> Strong in you, the Force is. He asked her to commit now or later.To commit now or later, he asked. Yes, hrrrm. Commit now or later, you must. The new girl was fired today at noon.Hrrmmm. Today at noon, the new"
```

# Underlying Cause

**TL;DR**: See [this bug report](https://github.com/huggingface/trl/issues/3140).

The `SFTTrainer` class is no longer correctly applying the chat template to the dataset. Since the dataset is no longer in chat template, it is missing all special tokens such as role tokens like `<|user|>` or `<|assistant|>` or end of sentence tokens like `<|endoftext|>`. This means there is no distinction between input and output and the LLM is treating the entire dataset as one continuous stream of text. This is what is leading the LLM to generate endless nonsense.

# Solution (Workaround)

1. Add the missing role and EOS tokens by converting dataset to conversational format.
```python
# Load the dataset
dataset = load_dataset("dvgodoy/yoda_sentences", split="train")

dataset = dataset.rename_column("sentence", "prompt")
dataset = dataset.rename_column("translation_extra", "completion")
dataset = dataset.remove_columns(["translation"])

# To fix missing role tokens, convert the dataset to conversational format.
def apply_chat_template(sample):
    sample['messages'] = [
        {"role":"user", "content":sample['prompt']},
        {"role":"assistant", "content":sample['completion']}]
    del sample['prompt']
    del sample['completion']
    return sample

dataset = dataset.map(apply_chat_template)
```

2. Prevent model from over-generating by removing `eos_token_id=tokenizer.eos_token_id` from `model.generate()`.
I don't know why this works, but now the model only generates _one_ response like it should.

```python
def generate(model, tokenizer, prompt, max_new_tokens=64, skip_special_tokens=False):
    tokenized_input = tokenizer(prompt, add_special_tokens=False, return_tensors="pt").to(model.device)

    model.eval()
    generation_output = model.generate(**tokenized_input,
                                       # Don't specify eos_token_id as a kwarg here.
                                       max_new_tokens=max_new_tokens)
    
    output = tokenizer.batch_decode(generation_output, 
                                    skip_special_tokens=skip_special_tokens)
    return output[0]
```

We now get the correct output from the LLM:
```python
"<|user|> The force is strong in you!<|end|><|assistant|> Strong in you, the force is!<|end|>"
```

**Comment by dvgodoy** [comment]
Hi @Alzter ,

Thank you for supporting my work and for this detailed report.

You're absolutely right, they made big changes to `trl` and the `instruction` format isn't properly supported anymore. Before, the `instructions_formatting_function()` would get called to properly convert the `prompt` and `completion` columns to the `conversational` format and then apply the template. Now, as you pointed out, they're simply concatenated.

I've updated the notebooks to include the conversion to the conversational format based on the code in `instructions_formatting_function()`.

There was yet another change that lead to the endless generation, as you also pointed out. Even though the dataset was packed, the collate function used was `DataCollatorWithPadding` (you can see it  as an attribute of the data loader returned `trainer.get_train_dataloader()`) in version 0.12.1. In version 0.15.2, however, the collate function is `DataCollatorForLanguageModeling` and this collator (unlike the previous one) masks padding tokens in `labels`.  

The issue is, Phi's tokenizer has the same token for both PAD and EOS, which makes the EOS token ignored as a valid label during training. That's why it simply won't stop generating new tokens - it has never seen the EOS token during training so it cannot generate it. The solution - as we're doing in a later chapter - is to assign the UNK to PAD token, keeping the EOS token unique and present in the label.

Once again, thanks for alerting me to this issue.

Best,
Daniel

**Comment by Alzter** [comment]
Since Daniel has implemented the workaround I'm going to mark this issue as closed :)
